pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/SegmentAssignment.java

```diff
@@ -52,6 +52,22 @@ public interface SegmentAssignment {
   List<String> assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
       Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap);
 
+  /**
+   * Assign new segment but check if the assignment calculated with instancePartition is consistent with the one set
+   * in the current idealState. If not, use the one from idealState to stay consistent with the source of truth.
+   *
+   * @param segmentName Name of the segment to be assigned
+   * @param currentAssignment Current segment assignment of the table (map from segment name to instance state map)
+   * @param instancePartitionsMap Map from type (OFFLINE|CONSUMING|COMPLETED) to instance partitions
+   * @param instancesAssigned returns the list of instances to assign the segment
+   * @return true if consistent or found no existing assignment from idealState; false otherwise.
+   */
+  default boolean assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
```

Comment:
I feel this is not a very clean abstraction. We can pass in an optional `ControllerMetrics` into the `SegmentAssignment`. We don't want to have one new method whenever we want to emit a metric

Reply:
I was thinking metrics was a bit heavy dep for SegmentAssignment interface (which seems quite functional), but I just saw its init() takes in HexliManager, so I'll pass a ControllerMetrics there too.